{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Kinase-based Viral Drug Connectivity Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intro\n",
    "Kinases are pretty exciting to me, not only because protein phosphorylation plays a role in regulating most aspects of cell life, but also because abnormal phosphorylation is often the cause or consequence of disease. Accordingly, they've become an important group of drug targets that I wanted to take a look at.\n",
    "\n",
    "In an attempt to unearth some computational biology skills, review some recent Pandas exposure, and get back into the spirit of biological research and discovery, I wanted to explore the relationship between kinase differential expression and viral infection. To do so, I ran a general query on the Gene Expression Atlas for all experiments that had to deal with viruses and for genes somehow related to kinases. Following this, the ultimate goal was to use connectivity mapping to find possible drug candidates correlated with a reversion in differential expression.\n",
    "\n",
    "## Some Background + Motivation\n",
    "Recently, one of five 2016 Breakthrough Prizes in the Life Sciences was awarded to Professor Stephen Elledge, who among other incredible things is particularly known for his work in eludicating the human DNA damage response. Excitingly, this can basically be summarized as a protein kinase cascade, and so one could imagine directing their drug discovery search in regards to diseases that heavily involve DNA damage towards kinase inhibitors or activators. Cancers are probably what come to mind the most, and I had also done some Pandas notebook exploration into that, but many cells also undergo virus infection-associated DDR. \n",
    "\n",
    "### Picking the right virus infection targets\n",
    "Another motivating idea is that going after viruses is difficult, particularly because of their rapid mutation cycle frequently rendering drugs obsolete. While we typically think of targeting foreign proteins such as external envelope proteins, another path to take might involve looking elsewhere from the virus entirely, and instead at host targets. While this sounds kind of like targeting your own cells with a therapeutic meant to kill viruses, viral infection might cause the upregulation of certain genes in host cells that allow them to propogate. If we can disrupt these then, the virus has a much harder time of getting by, and can't quite rely on its typical mutation frequency to escape dependency as fast. The idea is that drugs that disrupt host pathways that the virus relies on can't be as easily overcome by the virus through mutation because that would require the virus to evolve an entirely different functionality, which is a much lower probability event then changing a few coat protein residues.\n",
    "\n",
    "### More motivation\n",
    "\n",
    "Along with talking to Dr. Andrew Taube at DESRES and learning about their group's work in MD with kinases, phosphatases, and viruses, and having my own probably farfetched ideas about one day being able to engineer a sort of hyper DDR and immunity against the introduction of viral DNA, I was interested in seeing what I could find through a short exploration with publically available datasets and Pandas tools.  \n",
    "\n",
    "I had also just taken [MCB60: Cellular Biology and Molecular Medicine](https://locator.tlt.harvard.edu/course/colgsas-110424), which involved a lab component exploring DDR in yeast [(slides viewable here)](https://docs.google.com/presentation/d/1DE3JCsRNWs4jZbzqPJVzhuY_mkgSujVM9WoqNJsQork/edit?usp=sharing). Three months of intermittent web lab work is kind of naturally inconclusive, so I was eager to see if I could get some more definite results computationally. \n",
    "\n",
    "While in the short time putting this together I could only get as far as generating possibly connected drugs that would work towards reversing host cell expression after virus infection, and there's still a ton of analysis to be done with more time, the exercise was pretty fun and interesting. The later parts of the notebook could also definitely be cleaned up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data of all virus experiments with differential expression of kinase-associated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_diff_exp = pd.read_table('differentialResults-kinaseAndVirus.tsv')\n",
    "df_diff_exp_up = df_diff_exp.loc[df_diff_exp['log2foldchange'] >= 0]\n",
    "df_diff_exp_down = df_diff_exp.loc[df_diff_exp['log2foldchange'] < 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## View all experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       'PR8 influenza virus' vs 'none' in 'wild type'\n",
       "1    'Newcastle disease virus' at '10 hour' vs 'non...\n",
       "2    'Newcastle disease virus' at '12 hour' vs 'non...\n",
       "3    'Newcastle disease virus' at '10 hour' vs 'non...\n",
       "4    'Newcastle disease virus' at '14 hour' vs 'non...\n",
       "Name: Comparison, dtype: object"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diff_exp['Comparison'].drop_duplicates().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Isolating one experiment\n",
    "\n",
    "As a first workthrough example, we'll look at the 'PR8 influenza virus' vs 'none' in 'wild type' experiment and assign the data to the `df_diff_exp_0` dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Experiment Accession</th>\n",
       "      <th>Comparison</th>\n",
       "      <th>log2foldchange</th>\n",
       "      <th>pValue</th>\n",
       "      <th>tStatistics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000169245</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>9.9</td>\n",
       "      <td>4.069023e-10</td>\n",
       "      <td>64.458303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000271503</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>8.6</td>\n",
       "      <td>4.331436e-10</td>\n",
       "      <td>62.713264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSG00000134326</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>7.6</td>\n",
       "      <td>4.069023e-10</td>\n",
       "      <td>69.999440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENSG00000271503</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>7.4</td>\n",
       "      <td>4.346290e-08</td>\n",
       "      <td>31.419356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ENSG00000138646</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>7.2</td>\n",
       "      <td>4.069023e-10</td>\n",
       "      <td>65.716153</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene      Organism Experiment Accession  \\\n",
       "0   ENSG00000169245  homo sapiens         E-GEOD-13637   \n",
       "11  ENSG00000271503  homo sapiens         E-GEOD-13637   \n",
       "26  ENSG00000134326  homo sapiens         E-GEOD-13637   \n",
       "35  ENSG00000271503  homo sapiens         E-GEOD-13637   \n",
       "44  ENSG00000138646  homo sapiens         E-GEOD-13637   \n",
       "\n",
       "                                        Comparison  log2foldchange  \\\n",
       "0   'PR8 influenza virus' vs 'none' in 'wild type'             9.9   \n",
       "11  'PR8 influenza virus' vs 'none' in 'wild type'             8.6   \n",
       "26  'PR8 influenza virus' vs 'none' in 'wild type'             7.6   \n",
       "35  'PR8 influenza virus' vs 'none' in 'wild type'             7.4   \n",
       "44  'PR8 influenza virus' vs 'none' in 'wild type'             7.2   \n",
       "\n",
       "          pValue  tStatistics  \n",
       "0   4.069023e-10    64.458303  \n",
       "11  4.331436e-10    62.713264  \n",
       "26  4.069023e-10    69.999440  \n",
       "35  4.346290e-08    31.419356  \n",
       "44  4.069023e-10    65.716153  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diff_exp_0 = df_diff_exp.loc[df_diff_exp['Comparison']==df_diff_exp['Comparison'][0]]\n",
    "df_diff_exp_0.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Separating genes into up and down regulation dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_diff_exp_0_up = df_diff_exp_0.loc[df_diff_exp_0['log2foldchange'] > 0]\n",
    "df_diff_exp_0_down = df_diff_exp_0.loc[df_diff_exp_0['log2foldchange'] < 0]\n",
    "\n",
    "df_diff_exp_0_up['Experiment Accession'] = df_diff_exp_0['Experiment Accession']\n",
    "df_diff_exp_0_down['Experiment Accession'] = df_diff_exp_0['Experiment Accession']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Connecting to Biomart\n",
    "\n",
    "Connects to a pretty versatile server for annotating and querying for various components (CMAP tags, Ensembl gene names, etc.)\n",
    "\n",
    "Install module, define function to retrieve and query database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from biomart import BiomartServer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "bm_server = BiomartServer('http://www.ensembl.org/biomart')\n",
    "# bm_server.show_datasets()\n",
    "\n",
    "# Use the GRCh38.p7 human ensemble dataset\n",
    "hs_ensembl = bm_server.datasets['hsapiens_gene_ensembl']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining a get_probeset function\n",
    "Returns dataframe with Ensembl Gene ID and Affy HG U133A probeset from the Biomart Ensembl server \n",
    "* Requires a dataframe with a `Gene` column name specifying the Ensembl code, and the Biomart Ensembl server previously specified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_probeset(df, server):\n",
    "    response = server.search({\n",
    "            'filters': {\n",
    "                'with_affy_hg_u133a': True,\n",
    "                'ensembl_gene_id': df['Gene'].tolist()\n",
    "            },\n",
    "            'attributes': [\n",
    "                'ensembl_gene_id', 'affy_hg_u133a'\n",
    "            ]\n",
    "        })\n",
    "    try:\n",
    "        df_probeset = pd.DataFrame([line.decode('utf-8').split('\\t') for line in response.iter_lines()],\n",
    "                                   columns=['Gene ID', 'Affy HG U133A probeset'])\n",
    "        df_probeset['Experiment Accession'] = df['Experiment Accession']\n",
    "        return df_probeset\n",
    "    except:\n",
    "        # Inserts placeholder gene (involved in red-green color blindness) if gene set is empty\n",
    "        df_empty_probeset = pd.DataFrame({'Gene ID': ['ENSG00000102076'], 'Affy HG U133A probeset': ['221327_s_at']})\n",
    "        df_empty_probeset['Experiment Accession'] = df['Experiment Accession']\n",
    "        print('Empty dataset encountered. 0 probes returned')\n",
    "        return df_empty_probeset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting probe sets for first experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_probeset_0_up = get_probeset(df_diff_exp_0_up, hs_ensembl)\n",
    "df_probeset_0_down = get_probeset(df_diff_exp_0_down, hs_ensembl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert to .grp files for CMap querying\n",
    "The Broad Institute Connectivity Map generates gene expression signatures associated with perturbations by certain small molecule drugs. \n",
    "\n",
    "To query the service, the project builds signatures based on lists of HG-U133A probe sets associated with the up- and down-regulated genes stored in .grp files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function to do all of the above\n",
    "\n",
    "Generates .grp files ready for CMap querying\n",
    "\n",
    "*Note: make sure to have a directory named whatever you specify the `dirname` parameter in the same path level as the notebook directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def save_cmap_tags(df, server, filename, dirname, lim=0):\n",
    "    df_probeset = get_probeset(df, server)\n",
    "    if lim > 0:\n",
    "        df_probeset = df_probeset[:lim]\n",
    "    numtags = df_probeset.shape[0]\n",
    "    if numtags > 200:\n",
    "        print('Recommended to limit max number of tags to 200')\n",
    "    df_probeset['Affy HG U133A probeset'].drop_duplicates().to_csv(path=str(dirname)+'/'+str(filename)+'.grp', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "save_cmap_tags(df_diff_exp_0_up, hs_ensembl, 'probesets_exp_0_up', 'probesets')\n",
    "save_cmap_tags(df_diff_exp_0_down, hs_ensembl, 'probesets_exp_0_down', 'probesets')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Querying CMap\n",
    "The above function `save_cmap_tags` will save a `.grp` files to the notebook directory. These can then be used to build a query at [CMap](https://portals.broadinstitute.org/cmap/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to generate CMap query tags from Gene Expression Atlas upload\n",
    "\n",
    "__i.e. doing almost all of the above in one method for all experiments__\n",
    "\n",
    "Because automation is good.\n",
    "\n",
    "---\n",
    "\n",
    "Possibly helper functions are noted below. (Example follows afterward).\n",
    "\n",
    "*Note: make sure to have a directory named whatever you specify the `dirname` parameter in the same path level as the notebook directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def save_cmap_tags_from_atlas(df, reg, server, dirname, lim=0):\n",
    "    if reg == 0:\n",
    "        diff_id = '-down'\n",
    "    elif reg == 1:\n",
    "        diff_id = '-up'\n",
    "    df.groupby('Experiment Accession') \\\n",
    "        .apply(lambda df: save_cmap_tags(split_diff_exp(df, reg), server, str(df['Experiment Accession'] \\\n",
    "                                                                              .iloc[0])+diff_id, dirname)) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to isolate experiment by Experiment Accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_single_diff_exp(df, access_id):\n",
    "    df_single_diff_exp = df.loc[df['Experiment Accession'] == access_id]\n",
    "    return df_single_diff_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Experiment Accession</th>\n",
       "      <th>Comparison</th>\n",
       "      <th>log2foldchange</th>\n",
       "      <th>pValue</th>\n",
       "      <th>tStatistics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000169245</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>9.9</td>\n",
       "      <td>4.069023e-10</td>\n",
       "      <td>64.458303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000271503</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>8.6</td>\n",
       "      <td>4.331436e-10</td>\n",
       "      <td>62.713264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene      Organism Experiment Accession  \\\n",
       "0   ENSG00000169245  homo sapiens         E-GEOD-13637   \n",
       "11  ENSG00000271503  homo sapiens         E-GEOD-13637   \n",
       "\n",
       "                                        Comparison  log2foldchange  \\\n",
       "0   'PR8 influenza virus' vs 'none' in 'wild type'             9.9   \n",
       "11  'PR8 influenza virus' vs 'none' in 'wild type'             8.6   \n",
       "\n",
       "          pValue  tStatistics  \n",
       "0   4.069023e-10    64.458303  \n",
       "11  4.331436e-10    62.713264  "
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_single_diff_exp(df_diff_exp, 'E-GEOD-13637').head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to isolate data into up- and down-regulated gene sets\n",
    "Parameter __reg__ should be an integer 0 or 1 \n",
    "* 0 denotes downregulated genes\n",
    "* 1 denotes upregulated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def split_diff_exp(df, reg):\n",
    "    if reg == 0:\n",
    "        return df.loc[df['log2foldchange'] < 0]\n",
    "    elif reg == 1:\n",
    "        return df.loc[df['log2foldchange'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Experiment Accession</th>\n",
       "      <th>Comparison</th>\n",
       "      <th>log2foldchange</th>\n",
       "      <th>pValue</th>\n",
       "      <th>tStatistics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>ENSG00000137752</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>2.7</td>\n",
       "      <td>7.423240e-07</td>\n",
       "      <td>20.756294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>ENSG00000204264</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>2.7</td>\n",
       "      <td>7.830226e-06</td>\n",
       "      <td>15.050237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>ENSG00000121858</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'H5N1 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>2.7</td>\n",
       "      <td>3.086127e-04</td>\n",
       "      <td>8.774171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>ENSG00000047410</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'pEGZ-transfected (empty vector)' vs 'wild typ...</td>\n",
       "      <td>2.7</td>\n",
       "      <td>1.316403e-04</td>\n",
       "      <td>12.403622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>ENSG00000023445</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>E-GEOD-13637</td>\n",
       "      <td>'PR8 influenza virus' vs 'none' in 'wild type'</td>\n",
       "      <td>2.7</td>\n",
       "      <td>5.217151e-06</td>\n",
       "      <td>16.040361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene      Organism Experiment Accession  \\\n",
       "999  ENSG00000137752  homo sapiens         E-GEOD-13637   \n",
       "908  ENSG00000204264  homo sapiens         E-GEOD-13637   \n",
       "998  ENSG00000121858  homo sapiens         E-GEOD-13637   \n",
       "984  ENSG00000047410  homo sapiens         E-GEOD-13637   \n",
       "983  ENSG00000023445  homo sapiens         E-GEOD-13637   \n",
       "\n",
       "                                            Comparison  log2foldchange  \\\n",
       "999     'PR8 influenza virus' vs 'none' in 'wild type'             2.7   \n",
       "908     'PR8 influenza virus' vs 'none' in 'wild type'             2.7   \n",
       "998    'H5N1 influenza virus' vs 'none' in 'wild type'             2.7   \n",
       "984  'pEGZ-transfected (empty vector)' vs 'wild typ...             2.7   \n",
       "983     'PR8 influenza virus' vs 'none' in 'wild type'             2.7   \n",
       "\n",
       "           pValue  tStatistics  \n",
       "999  7.423240e-07    20.756294  \n",
       "908  7.830226e-06    15.050237  \n",
       "998  3.086127e-04     8.774171  \n",
       "984  1.316403e-04    12.403622  \n",
       "983  5.217151e-06    16.040361  "
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example building on above. Note the return of all positive log2 fold changes\n",
    "split_diff_exp(get_single_diff_exp(df_diff_exp, 'E-GEOD-13637'), 1).sort_values('log2foldchange', ascending=True).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving CMap tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty dataset encountered. 0 probes returned\n",
      "Empty dataset encountered. 0 probes returned\n",
      "Empty dataset encountered. 0 probes returned\n",
      "Empty dataset encountered. 0 probes returned\n",
      "Empty dataset encountered. 0 probes returned\n",
      "Empty dataset encountered. 0 probes returned\n",
      "Empty dataset encountered. 0 probes returned\n"
     ]
    }
   ],
   "source": [
    "# Hacky way to get both down- and up-regulated tag probe sets\n",
    "for i in range(0, 2):\n",
    "    save_cmap_tags_from_atlas(df=df_diff_exp, reg=i, server=hs_ensembl, dirname='probesets')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The result:\n",
    "<img src=\"kinase_virus-probes.png\">\n",
    "\n",
    "*Note: last 2 files are artifacts of the example workthrough.\n",
    "\n",
    "## Analysis of connectivity with CMAP\n",
    "We now have `.grp` probe set files that can be used to generate a signature and query for CMAP. After doing so, we can download the resulting table to identify perturbations that have antagonistic gene expression effects denoted by negative enrichment scores (which could be good for identifying possible drugs)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example Analysis with E-GEOD-8717\n",
    "\n",
    "### Querying CMap with .grp files\n",
    "<img src=\"cmap_results-a_E-GEOD-8717.png\" style=\"width: 500px;\">\n",
    "\n",
    "### Query complete!\n",
    "<img src=\"cmap_results-b_E-GEOD-8717.png\" style=\"width: 500px;\">\n",
    "\n",
    "### Loading the connectivity mapping dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank</th>\n",
       "      <th>cmap name</th>\n",
       "      <th>mean</th>\n",
       "      <th>n</th>\n",
       "      <th>enrichment</th>\n",
       "      <th>p</th>\n",
       "      <th>specificity</th>\n",
       "      <th>percent non-null</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ouabain</td>\n",
       "      <td>0.777</td>\n",
       "      <td>4</td>\n",
       "      <td>0.979</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>digoxin</td>\n",
       "      <td>0.778</td>\n",
       "      <td>4</td>\n",
       "      <td>0.978</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0047</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>emetine</td>\n",
       "      <td>0.749</td>\n",
       "      <td>4</td>\n",
       "      <td>0.955</td>\n",
       "      <td>0</td>\n",
       "      <td>0.007</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>anisomycin</td>\n",
       "      <td>0.748</td>\n",
       "      <td>4</td>\n",
       "      <td>0.947</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0412</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>cicloheximide</td>\n",
       "      <td>0.652</td>\n",
       "      <td>4</td>\n",
       "      <td>0.941</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rank      cmap name   mean  n  enrichment  p specificity  percent non-null\n",
       "0     1        ouabain  0.777  4       0.979  0      0.0051               100\n",
       "1     2        digoxin  0.778  4       0.978  0      0.0047               100\n",
       "2     3        emetine  0.749  4       0.955  0       0.007               100\n",
       "3     4     anisomycin  0.748  4       0.947  0      0.0412               100\n",
       "4     5  cicloheximide  0.652  4       0.941  0      0.0226               100"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_E_GEOD_8717 = pd.read_excel('permutedResults-E-GEOD-8717.xls')\n",
    "df_E_GEOD_8717.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying possible connected drug candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank</th>\n",
       "      <th>cmap name</th>\n",
       "      <th>mean</th>\n",
       "      <th>n</th>\n",
       "      <th>enrichment</th>\n",
       "      <th>p</th>\n",
       "      <th>specificity</th>\n",
       "      <th>percent non-null</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>naloxone</td>\n",
       "      <td>-0.379</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.680</td>\n",
       "      <td>0.00288</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>38</td>\n",
       "      <td>talampicillin</td>\n",
       "      <td>-0.301</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.748</td>\n",
       "      <td>0.00802</td>\n",
       "      <td>0.0214</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>39</td>\n",
       "      <td>phensuximide</td>\n",
       "      <td>-0.433</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.748</td>\n",
       "      <td>0.00808</td>\n",
       "      <td>0.0191</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>45</td>\n",
       "      <td>NU-1025</td>\n",
       "      <td>-0.612</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.926</td>\n",
       "      <td>0.01131</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>46</td>\n",
       "      <td>ganciclovir</td>\n",
       "      <td>-0.275</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.721</td>\n",
       "      <td>0.01229</td>\n",
       "      <td>0.0113</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rank      cmap name   mean  n  enrichment        p specificity  \\\n",
       "24    25       naloxone -0.379  6      -0.680  0.00288      0.0155   \n",
       "37    38  talampicillin -0.301  4      -0.748  0.00802      0.0214   \n",
       "38    39   phensuximide -0.433  4      -0.748  0.00808      0.0191   \n",
       "44    45        NU-1025 -0.612  2      -0.926  0.01131      0.0107   \n",
       "45    46    ganciclovir -0.275  4      -0.721  0.01229      0.0113   \n",
       "\n",
       "    percent non-null  \n",
       "24                66  \n",
       "37                50  \n",
       "38                75  \n",
       "44               100  \n",
       "45                50  "
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_candidates(df):\n",
    "    df_cand = df.loc[df['enrichment'] < 0].sort_values(['rank', 'enrichment'], ascending=True)\n",
    "    return df_cand\n",
    "\n",
    "find_candidates(df_E_GEOD_8717).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading all CMap dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_E_GEOD_13637 = pd.read_excel('permutedResults-E-GEOD-13637.xls')\n",
    "df_E_GEOD_17156 = pd.read_excel('permutedResults-E-GEOD-17156.xls')\n",
    "df_E_GEOD_17400 = pd.read_excel('permutedResults-E-GEOD-17400.xls')\n",
    "df_E_GEOD_18791 = pd.read_excel('permutedResults-E-GEOD-18791.xls')\n",
    "df_E_GEOD_18816 = pd.read_excel('permutedResults-E-GEOD-18816.xls')\n",
    "df_E_GEOD_19665 = pd.read_excel('permutedResults-E-GEOD-19665.xls')\n",
    "df_E_GEOD_22522 = pd.read_excel('permutedResults-E-GEOD-22522.xls')\n",
    "df_E_GEOD_31193 = pd.read_excel('permutedResults-E-GEOD-31193.xls')\n",
    "df_E_GEOD_37571 = pd.read_excel('permutedResults-E-GEOD-37571.xls')\n",
    "df_E_GEOD_40844 = pd.read_excel('permutedResults-E-GEOD-40844.xls')\n",
    "df_E_MEXP_1274 = pd.read_excel('permutedResults-E-MEXP-1274.xls')\n",
    "df_E_MEXP_3582 = pd.read_excel('permutedResults-E-MEXP-3582.xls')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank</th>\n",
       "      <th>cmap name</th>\n",
       "      <th>mean</th>\n",
       "      <th>n</th>\n",
       "      <th>enrichment</th>\n",
       "      <th>p</th>\n",
       "      <th>specificity</th>\n",
       "      <th>percent non-null</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>H-7</td>\n",
       "      <td>-0.713</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.944</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0909</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>neostigmine bromide</td>\n",
       "      <td>-0.574</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.831</td>\n",
       "      <td>0.00147</td>\n",
       "      <td>0.0115</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>sulfamethizole</td>\n",
       "      <td>-0.728</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.821</td>\n",
       "      <td>0.00193</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>bambuterol</td>\n",
       "      <td>-0.530</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.795</td>\n",
       "      <td>0.00348</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>mexiletine</td>\n",
       "      <td>-0.302</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.652</td>\n",
       "      <td>0.00481</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>decamethonium bromide</td>\n",
       "      <td>-0.318</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.775</td>\n",
       "      <td>0.00527</td>\n",
       "      <td>0.0059</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>lisinopril</td>\n",
       "      <td>-0.657</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.854</td>\n",
       "      <td>0.00619</td>\n",
       "      <td>0.0216</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>bufexamac</td>\n",
       "      <td>-0.299</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.739</td>\n",
       "      <td>0.00905</td>\n",
       "      <td>0.0177</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>metolazone</td>\n",
       "      <td>-0.492</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.670</td>\n",
       "      <td>0.00907</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>diltiazem</td>\n",
       "      <td>-0.398</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.670</td>\n",
       "      <td>0.00907</td>\n",
       "      <td>0.0313</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>carbenoxolone</td>\n",
       "      <td>-0.397</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.736</td>\n",
       "      <td>0.00975</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>29</td>\n",
       "      <td>papaverine</td>\n",
       "      <td>-0.470</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.729</td>\n",
       "      <td>0.01104</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>30</td>\n",
       "      <td>pivmecillinam</td>\n",
       "      <td>-0.267</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.728</td>\n",
       "      <td>0.01118</td>\n",
       "      <td>0.0221</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>31</td>\n",
       "      <td>epitiostanol</td>\n",
       "      <td>-0.404</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.726</td>\n",
       "      <td>0.01154</td>\n",
       "      <td>0.037</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>35</td>\n",
       "      <td>alprenolol</td>\n",
       "      <td>-0.417</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.713</td>\n",
       "      <td>0.01373</td>\n",
       "      <td>0.01</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>36</td>\n",
       "      <td>metaraminol</td>\n",
       "      <td>-0.504</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.702</td>\n",
       "      <td>0.01643</td>\n",
       "      <td>0.0112</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>38</td>\n",
       "      <td>lymecycline</td>\n",
       "      <td>-0.469</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.694</td>\n",
       "      <td>0.01846</td>\n",
       "      <td>0.0352</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>39</td>\n",
       "      <td>dilazep</td>\n",
       "      <td>-0.458</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.627</td>\n",
       "      <td>0.01868</td>\n",
       "      <td>0.0345</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>43</td>\n",
       "      <td>ethotoin</td>\n",
       "      <td>-0.291</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.573</td>\n",
       "      <td>0.02161</td>\n",
       "      <td>0.1414</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>44</td>\n",
       "      <td>riluzole</td>\n",
       "      <td>-0.415</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.616</td>\n",
       "      <td>0.02289</td>\n",
       "      <td>0.019</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>45</td>\n",
       "      <td>alfaxalone</td>\n",
       "      <td>-0.487</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.776</td>\n",
       "      <td>0.02299</td>\n",
       "      <td>0.0319</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>49</td>\n",
       "      <td>raubasine</td>\n",
       "      <td>-0.317</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.670</td>\n",
       "      <td>0.02638</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>51</td>\n",
       "      <td>canavanine</td>\n",
       "      <td>-0.351</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.756</td>\n",
       "      <td>0.02981</td>\n",
       "      <td>0.0412</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>52</td>\n",
       "      <td>2,6-dimethylpiperidine</td>\n",
       "      <td>-0.297</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.597</td>\n",
       "      <td>0.03038</td>\n",
       "      <td>0.0108</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>53</td>\n",
       "      <td>paroxetine</td>\n",
       "      <td>-0.319</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.660</td>\n",
       "      <td>0.03044</td>\n",
       "      <td>0.0398</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>55</td>\n",
       "      <td>oxprenolol</td>\n",
       "      <td>-0.497</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.656</td>\n",
       "      <td>0.03195</td>\n",
       "      <td>0.0396</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>56</td>\n",
       "      <td>sitosterol</td>\n",
       "      <td>-0.532</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.656</td>\n",
       "      <td>0.03213</td>\n",
       "      <td>0.1071</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>57</td>\n",
       "      <td>labetalol</td>\n",
       "      <td>-0.534</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.655</td>\n",
       "      <td>0.03251</td>\n",
       "      <td>0.036</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>59</td>\n",
       "      <td>eldeline</td>\n",
       "      <td>-0.305</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.543</td>\n",
       "      <td>0.03613</td>\n",
       "      <td>0.0756</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>68</td>\n",
       "      <td>nicotinic acid</td>\n",
       "      <td>-0.259</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.637</td>\n",
       "      <td>0.04156</td>\n",
       "      <td>0.0599</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1261</th>\n",
       "      <td>1262</td>\n",
       "      <td>buflomedil</td>\n",
       "      <td>0.130</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.213</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1263</th>\n",
       "      <td>1264</td>\n",
       "      <td>chloroquine</td>\n",
       "      <td>-0.044</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.211</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266</th>\n",
       "      <td>1267</td>\n",
       "      <td>rosiglitazone</td>\n",
       "      <td>0.036</td>\n",
       "      <td>14</td>\n",
       "      <td>-0.208</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1268</th>\n",
       "      <td>1269</td>\n",
       "      <td>(-)-isoprenaline</td>\n",
       "      <td>-0.016</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.206</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1270</th>\n",
       "      <td>1271</td>\n",
       "      <td>nifedipine</td>\n",
       "      <td>-0.099</td>\n",
       "      <td>7</td>\n",
       "      <td>-0.200</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1271</th>\n",
       "      <td>1272</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>0.082</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.200</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1273</th>\n",
       "      <td>1274</td>\n",
       "      <td>tretinoin</td>\n",
       "      <td>-0.110</td>\n",
       "      <td>22</td>\n",
       "      <td>-0.199</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1274</th>\n",
       "      <td>1275</td>\n",
       "      <td>bucladesine</td>\n",
       "      <td>0.009</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.198</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>1276</td>\n",
       "      <td>mafenide</td>\n",
       "      <td>-0.013</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.197</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276</th>\n",
       "      <td>1277</td>\n",
       "      <td>pregnenolone</td>\n",
       "      <td>0.023</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.197</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1277</th>\n",
       "      <td>1278</td>\n",
       "      <td>dicoumarol</td>\n",
       "      <td>-0.035</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.194</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279</th>\n",
       "      <td>1280</td>\n",
       "      <td>naphazoline</td>\n",
       "      <td>0.004</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.193</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1280</th>\n",
       "      <td>1281</td>\n",
       "      <td>gliquidone</td>\n",
       "      <td>-0.044</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.192</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1281</th>\n",
       "      <td>1282</td>\n",
       "      <td>droperidol</td>\n",
       "      <td>-0.001</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.191</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1284</th>\n",
       "      <td>1285</td>\n",
       "      <td>chlorpromazine</td>\n",
       "      <td>-0.073</td>\n",
       "      <td>19</td>\n",
       "      <td>-0.190</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1286</th>\n",
       "      <td>1287</td>\n",
       "      <td>acetylsalicylic acid</td>\n",
       "      <td>-0.058</td>\n",
       "      <td>13</td>\n",
       "      <td>-0.190</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1290</th>\n",
       "      <td>1291</td>\n",
       "      <td>Prestwick-1080</td>\n",
       "      <td>-0.044</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.182</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1291</th>\n",
       "      <td>1292</td>\n",
       "      <td>Prestwick-664</td>\n",
       "      <td>-0.037</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.180</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1292</th>\n",
       "      <td>1293</td>\n",
       "      <td>estradiol</td>\n",
       "      <td>0.025</td>\n",
       "      <td>37</td>\n",
       "      <td>-0.180</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1293</th>\n",
       "      <td>1294</td>\n",
       "      <td>monastrol</td>\n",
       "      <td>0.067</td>\n",
       "      <td>8</td>\n",
       "      <td>-0.179</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1294</th>\n",
       "      <td>1295</td>\n",
       "      <td>naproxen</td>\n",
       "      <td>0.063</td>\n",
       "      <td>9</td>\n",
       "      <td>-0.179</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1295</th>\n",
       "      <td>1296</td>\n",
       "      <td>cefamandole</td>\n",
       "      <td>-0.018</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.178</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297</th>\n",
       "      <td>1298</td>\n",
       "      <td>wortmannin</td>\n",
       "      <td>0.056</td>\n",
       "      <td>18</td>\n",
       "      <td>-0.174</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1299</th>\n",
       "      <td>1300</td>\n",
       "      <td>lomefloxacin</td>\n",
       "      <td>-0.022</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.169</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300</th>\n",
       "      <td>1301</td>\n",
       "      <td>valproic acid</td>\n",
       "      <td>0.061</td>\n",
       "      <td>57</td>\n",
       "      <td>-0.164</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1303</th>\n",
       "      <td>1304</td>\n",
       "      <td>LY-294002</td>\n",
       "      <td>-0.064</td>\n",
       "      <td>61</td>\n",
       "      <td>-0.149</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1304</th>\n",
       "      <td>1305</td>\n",
       "      <td>fulvestrant</td>\n",
       "      <td>-0.104</td>\n",
       "      <td>40</td>\n",
       "      <td>-0.148</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1306</th>\n",
       "      <td>1307</td>\n",
       "      <td>tanespimycin</td>\n",
       "      <td>-0.047</td>\n",
       "      <td>62</td>\n",
       "      <td>-0.135</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1307</th>\n",
       "      <td>1308</td>\n",
       "      <td>haloperidol</td>\n",
       "      <td>-0.025</td>\n",
       "      <td>32</td>\n",
       "      <td>-0.131</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308</th>\n",
       "      <td>1309</td>\n",
       "      <td>sirolimus</td>\n",
       "      <td>-0.059</td>\n",
       "      <td>44</td>\n",
       "      <td>-0.113</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>709 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      rank               cmap name   mean   n  enrichment        p  \\\n",
       "1        2                     H-7 -0.713   4      -0.944        0   \n",
       "10      11     neostigmine bromide -0.574   4      -0.831  0.00147   \n",
       "11      12          sulfamethizole -0.728   4      -0.821  0.00193   \n",
       "13      14              bambuterol -0.530   4      -0.795  0.00348   \n",
       "14      15              mexiletine -0.302   6      -0.652  0.00481   \n",
       "15      16   decamethonium bromide -0.318   4      -0.775  0.00527   \n",
       "17      18              lisinopril -0.657   3      -0.854  0.00619   \n",
       "22      23               bufexamac -0.299   4      -0.739  0.00905   \n",
       "23      24              metolazone -0.492   5      -0.670  0.00907   \n",
       "24      25               diltiazem -0.398   5      -0.670  0.00907   \n",
       "25      26           carbenoxolone -0.397   4      -0.736  0.00975   \n",
       "28      29              papaverine -0.470   4      -0.729  0.01104   \n",
       "29      30           pivmecillinam -0.267   4      -0.728  0.01118   \n",
       "30      31            epitiostanol -0.404   4      -0.726  0.01154   \n",
       "34      35              alprenolol -0.417   4      -0.713  0.01373   \n",
       "35      36             metaraminol -0.504   4      -0.702  0.01643   \n",
       "37      38             lymecycline -0.469   4      -0.694  0.01846   \n",
       "38      39                 dilazep -0.458   5      -0.627  0.01868   \n",
       "42      43                ethotoin -0.291   6      -0.573  0.02161   \n",
       "43      44                riluzole -0.415   5      -0.616  0.02289   \n",
       "44      45              alfaxalone -0.487   3      -0.776  0.02299   \n",
       "48      49               raubasine -0.317   4      -0.670  0.02638   \n",
       "50      51              canavanine -0.351   3      -0.756  0.02981   \n",
       "51      52  2,6-dimethylpiperidine -0.297   5      -0.597  0.03038   \n",
       "52      53              paroxetine -0.319   4      -0.660  0.03044   \n",
       "54      55              oxprenolol -0.497   4      -0.656  0.03195   \n",
       "55      56              sitosterol -0.532   4      -0.656  0.03213   \n",
       "56      57               labetalol -0.534   4      -0.655  0.03251   \n",
       "58      59                eldeline -0.305   6      -0.543  0.03613   \n",
       "67      68          nicotinic acid -0.259   4      -0.637  0.04156   \n",
       "...    ...                     ...    ...  ..         ...      ...   \n",
       "1261  1262              buflomedil  0.130   4      -0.213      ---   \n",
       "1263  1264             chloroquine -0.044   4      -0.211      ---   \n",
       "1266  1267           rosiglitazone  0.036  14      -0.208      ---   \n",
       "1268  1269        (-)-isoprenaline -0.016   4      -0.206      ---   \n",
       "1270  1271              nifedipine -0.099   7      -0.200      ---   \n",
       "1271  1272             gemfibrozil  0.082   5      -0.200      ---   \n",
       "1273  1274               tretinoin -0.110  22      -0.199      ---   \n",
       "1274  1275             bucladesine  0.009   6      -0.198      ---   \n",
       "1275  1276                mafenide -0.013   5      -0.197      ---   \n",
       "1276  1277            pregnenolone  0.023   4      -0.197      ---   \n",
       "1277  1278              dicoumarol -0.035   6      -0.194      ---   \n",
       "1279  1280             naphazoline  0.004   5      -0.193      ---   \n",
       "1280  1281              gliquidone -0.044   4      -0.192      ---   \n",
       "1281  1282              droperidol -0.001   4      -0.191      ---   \n",
       "1284  1285          chlorpromazine -0.073  19      -0.190      ---   \n",
       "1286  1287    acetylsalicylic acid -0.058  13      -0.190      ---   \n",
       "1290  1291          Prestwick-1080 -0.044   4      -0.182      ---   \n",
       "1291  1292           Prestwick-664 -0.037   6      -0.180      ---   \n",
       "1292  1293               estradiol  0.025  37      -0.180      ---   \n",
       "1293  1294               monastrol  0.067   8      -0.179      ---   \n",
       "1294  1295                naproxen  0.063   9      -0.179      ---   \n",
       "1295  1296             cefamandole -0.018   4      -0.178      ---   \n",
       "1297  1298              wortmannin  0.056  18      -0.174      ---   \n",
       "1299  1300            lomefloxacin -0.022   6      -0.169      ---   \n",
       "1300  1301           valproic acid  0.061  57      -0.164      ---   \n",
       "1303  1304               LY-294002 -0.064  61      -0.149      ---   \n",
       "1304  1305             fulvestrant -0.104  40      -0.148      ---   \n",
       "1306  1307            tanespimycin -0.047  62      -0.135      ---   \n",
       "1307  1308             haloperidol -0.025  32      -0.131      ---   \n",
       "1308  1309               sirolimus -0.059  44      -0.113      ---   \n",
       "\n",
       "     specificity  percent non-null  \n",
       "1         0.0909               100  \n",
       "10        0.0115               100  \n",
       "11             0               100  \n",
       "13             0                75  \n",
       "14             0                50  \n",
       "15        0.0059                50  \n",
       "17        0.0216               100  \n",
       "22        0.0177                50  \n",
       "23             0                80  \n",
       "24        0.0313                60  \n",
       "25        0.0126                75  \n",
       "28        0.0126                75  \n",
       "29        0.0221                50  \n",
       "30         0.037                75  \n",
       "34          0.01                75  \n",
       "35        0.0112                75  \n",
       "37        0.0352                75  \n",
       "38        0.0345                80  \n",
       "42        0.1414                66  \n",
       "43         0.019                80  \n",
       "44        0.0319                66  \n",
       "48             0                50  \n",
       "50        0.0412                66  \n",
       "51        0.0108                60  \n",
       "52        0.0398                50  \n",
       "54        0.0396                75  \n",
       "55        0.1071                75  \n",
       "56         0.036                75  \n",
       "58        0.0756                50  \n",
       "67        0.0599                50  \n",
       "...          ...               ...  \n",
       "1261         ---                 0  \n",
       "1263         ---                25  \n",
       "1266         ---                14  \n",
       "1268         ---                25  \n",
       "1270         ---                28  \n",
       "1271         ---                20  \n",
       "1273         ---                31  \n",
       "1274         ---                16  \n",
       "1275         ---                20  \n",
       "1276         ---                25  \n",
       "1277         ---                16  \n",
       "1279         ---                20  \n",
       "1280         ---                25  \n",
       "1281         ---                25  \n",
       "1284         ---                15  \n",
       "1286         ---                15  \n",
       "1290         ---                25  \n",
       "1291         ---                16  \n",
       "1292         ---                21  \n",
       "1293         ---                12  \n",
       "1294         ---                11  \n",
       "1295         ---                25  \n",
       "1297         ---                22  \n",
       "1299         ---                33  \n",
       "1300         ---                12  \n",
       "1303         ---                26  \n",
       "1304         ---                30  \n",
       "1306         ---                19  \n",
       "1307         ---                18  \n",
       "1308         ---                22  \n",
       "\n",
       "[709 rows x 8 columns]"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_candidates(df_E_GEOD_13637)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
